app/services/events/service.py

```diff
@@ -0,0 +1,177 @@
+            end_date: End date for event search
+            categories: List of event categories to search for
+
+        Returns:
+            List of Event objects (limited to 10 results)
+
+        Raises:
+            ValidationError: If input parameters are invalid (from Pydantic)
+            ProviderError: If the event provider fails
+        """
+        logger.info(
+            f"Getting events for city: {city}, date range: {start_date} to {end_date}, categories: {categories}"
+        )
+
+        try:
+            # Validate input using Pydantic model
+            request = EventRequest(
+                city=city,
+                start_date=start_date,
+                end_date=end_date,
+                categories=categories,
+            )
+
+            # Check cache first
+            cache_key = f"{request.city}-{request.start_date}-{request.end_date}-{request.categories or 'all'}"
+            if self.cache:
+                cached_events = self.cache.get(cache_key)
+                if cached_events:
+                    return cached_events
+                logger.debug("Cache miss: fetching from provider")
+
+            # Fetch events from provider
+            logger.debug("Fetching events from provider")
+            events = self.provider.fetch(
+                request.city, request.start_date, request.end_date, request.categories
+            )
+            logger.info(f"Provider returned {len(events)} events")
+
+            # Limit results to 10
+            limited_events = events[:10]
+            logger.info(f"Returning {len(limited_events)} events (limited to 10)")
+
+            # Cache the results
+            if self.cache:
+                logger.debug("Caching results")
+                self.cache.set(cache_key, limited_events)
+
+            return limited_events
+
+        except Exception as e:
+            logger.error(f"Failed to get events: {e}")
+            raise
+
+    def _map_events_with_day_and_timeslot(
+        self, events: List[Event], start_date: datetime.date
+    ) -> Dict[str, dict]:
+        """
+        Map events to itinerary in the format:
+        {
+          "Day 1": { "afternoon": [ {...}, {...} ] },
+          "Day 2": { "evening": [ {...} ] }
+        }
+
+        - Uses "Day N" based on difference from `start_date`
+        - Skips empty blocks (only include blocks with events)
+        """
+        mapped = {}
+
+        for e in events:
+            event_date = e.date.date()
+            day_index = (event_date - start_date).days + 1
+            day_key = f"Day {day_index}"
+            hour = e.date.hour
+            if 8 <= hour < 12:
+                block = "morning"
+            elif 12 <= hour < 17:
+                block = "afternoon"
+            elif 17 <= hour < 23:
+                block = "evening"
+            else:
+                # Skip events outside itinerary range (night, early morning, etc.)
+                continue
```

Comment:
The block assignment for time-of-day is hardcoded; refactor into a helper function for reusability and clarity. E.g.:
```
def get_time_block(hour):
    if 8 <= hour < 12:
        return "morning"
    elif 12 <= hour < 17:
        return "afternoon"
    elif 17 <= hour < 23:
        return "evening"
    return None
block = get_time_block(hour)
if not block:
    continue
```